e3sm_to_cmip/util.py

```diff
@@ -54,6 +60,61 @@ def print_debug(e):
     traceback.print_tb(tb)
     print(e)
 
+def setup_logging(loglevel, logpath):
+    logname = logpath + "-" + UTC.localize(datetime.utcnow()).strftime("%Y%m%d_%H%M%S_%f")
+    if loglevel == "debug":
+        level = logging.DEBUG
+    elif loglevel == "error":
+        level = logging.ERROR
+    elif loglevel == "warning":
+        level = logging.WARNING
+    else:
+        level = logging.INFO
+    logging.basicConfig(
+        filename=logname,
+        # format="%(asctime)s:%(levelname)s:%(module)s:%(message)s",
+        format="%(asctime)s_%(msecs)03d:%(levelname)s:%(message)s",
+        datefmt="%Y%m%d_%H%M%S",
+        level=level,
+    )
+    logging.Formatter.converter = time.gmtime
+    # should be a separate message call
+    # logging.info(f"Starting up the warehouse with parameters: \n{pformat(self.__dict__)}")
+
+def log_message(level, message, user_level='INFO'):  # message BOTH to log file and to console (in color)
+
+    process_stack = inspect.stack()[1]
+    parent_module = inspect.getmodule(process_stack[0])
+
+    parent_name = parent_module.__name__.split(".")[-1].upper()
+    if parent_name == "__MAIN__":
+        parent_name = process_stack[1].split(".")[0].upper()
+    message = f"{parent_name}:{message}"
+
+    level = level.upper()
+    colors = {"INFO": "white", "WARNING": "yellow", "ERROR": "red", "DEBUG": "cyan"}
+    color = colors.get(level, 'red')
+    tstamp = UTC.localize(datetime.utcnow()).strftime("%Y%m%d_%H%M%S_%f")  # for console output
+    # first, print to logfile
+    if level == "DEBUG":
+        logging.debug(message)
+    elif level == "ERROR":
+        logging.error(message)
+    elif level == "WARNING":
+        logging.warning(message)
+    elif level == "INFO":
+        logging.info(message)
+    else:
+        print(f"ERROR: {level} is not a valid log level")
+
+    if level == 'DEBUG' and user_level != level:
+        pass
+    else:
+        # now to the console
+        msg = f"{tstamp}:{level}:{message}"
+        cprint(msg, color)
```

Comment:
I would recommend to just print the logger message, instead of colorizing them because it adds a new dependency (`termcolor`).

Reply:
Refer to my alternative approach in the main review comment though.

Reply:
Any format that supports sub-second timestamps (with only underscores as punctuation) and automatically includes the calling module name is fine by me.  I don't really care for termcolor.  Think of my supplied "log_message()" function as a mere suggestion.  I'm not sure I follow how to implement the alternative solution.  I'll search through the examples you've given.

Reply:
You can customize the log message format here: https://github.com/xCDAT/xcdat/blob/4a35d6a6131fe3fec22593f54a9e48b640ceac4f/xcdat/logger.py#L46

Reply:
I created a test-harness for the logger in the example, and then called it with

> log.info("Hello World")

The output produced was:

> 2022-05-13 10:37:09,743 [INFO]: test_logger2.py(main:62) >> Hello World

I far prefer:

> 20220513_103709_734:INFO:MAIN:Hello World

The example output is a zoo of delimiter characters (hyphens, whitespace, colons, commas, square-braces, periods, parentheses, and '>>'.  Somehow, they overlooked ~ ` @ # $ % ^ & * { } and |.

Also, my "setup_logging" includes an output file to capture the log output.

I will see about modifying the format.   A "smarter" logger would allow the addition of "func_name", and "line_no" for debug or errors, and not for "info", etc.

Reply:
The colors stuff is gone, and the system is simplified.  I think I'd like to apply what I've learned here to datasm - but that will be a major effort in translation.